addons/binding/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/ControllerHandler.java

```diff
@@ -0,0 +1,234 @@
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.math.BigDecimal;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.net.UnknownHostException;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseBridgeHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.openhab.binding.coolmasternet.handler.HVACHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/* Bridge to access a CoolMasterNet unit'sASCII protocol via TCP socket.
+ *
+ * A single CoolMasterNet can be connected to one or more HVAC units, each with a unique UID.
+ * These are individual Things inside the bridge.
+ */
+public class ControllerHandler extends BaseBridgeHandler {
+    private static final int SOCKET_TIMEOUT = 2000;
+    private static final Logger logger = LoggerFactory.getLogger(ControllerHandler.class);
+    private String host;
+    private int port;
+    private Socket socket;
+    private final Object lock = new Object();
+    private ScheduledFuture<?> refreshJob;
+
+    public ControllerHandler(Bridge thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initialising CoolMasterNet Controller handler...");
+        super.initialize();
+
+        Configuration config = this.getConfig();
+        host = (String) config.get(HOST);
+        port = 10102;
+        try {
+            port = ((BigDecimal) config.get(PORT)).intValue();
+        } catch (NullPointerException e) {
+            // keep default
+        }
+
+        int refresh = 5;
+        try {
+            refresh = ((BigDecimal) config.get(REFRESH)).intValue();
+        } catch (NullPointerException e) {
+            // keep default
+        }
+
+        Runnable refreshHVACUnits = new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    checkConnection();
+                    updateStatus(ThingStatus.ONLINE);
+                    for (Thing t : getThing().getThings()) {
+                        HVACHandler h = (HVACHandler) t.getHandler();
+                        h.refresh();
+                    }
+                } catch (CoolMasterClientError e) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            String.format("Could not connect to CoolMasterNet: {}", e.getMessage()));
```

Comment:
e.getMessage() should be good enough - better keep the string short to give the UI a chance to fit it on the screen.